[JIT] Disallow plain Dict type annotation without arg

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44586 [JIT] Disallow plain Optional type annotation without arg
* #44585 [JIT] Disallow plain Tuple type annotation without arg
* #44584 [JIT] Disallow plain List type annotation without arg
* **#44334 [JIT] Disallow plain Dict type annotation without arg**

**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

Differential Revision: [D23610766](https://our.internmc.facebook.com/intern/diff/D23610766)

Commit:
[JIT] Disallow plain Dict type annotation without arg

**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

[ghstack-poisoned]

gmagogsfm commented on test/jit/test_list_dict.py:
```diff
@@ -1514,3 +1514,14 @@ def test_dict_error():
         with self.assertRaisesRegex(Exception, "Arguments for call are not"):
             torch.jit.script(test_dict_error)
+
+    def test_no_type_annotation(self):
```
Nit:
test_type_annotation_missing_contained_type

SplitInfinity replied:
Fixed.

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -291,7 +291,8 @@ def try_ann_to_type(ann, loc):
         key = try_ann_to_type(ann.__args__[0], loc)
         value = try_ann_to_type(ann.__args__[1], loc)
-        return DictType(key, value)
+        if key and value:
```
What do you think about generate error message here? IMHO, "unknown type name" might not be clear enough to users since Dict is a valid type.

SplitInfinity replied:
I added warnings let me know what you think.

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

[ghstack-poisoned]

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

[ghstack-poisoned]

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

[ghstack-poisoned]

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

Differential Revision: [D23610766](https://our.internmc.facebook.com/intern/diff/D23610766)

[ghstack-poisoned]

eellison commented:
Another type container i missed: `Tuple` - could you check that this works for that too?

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

Differential Revision: [D23610766](https://our.internmc.facebook.com/intern/diff/D23610766)

[ghstack-poisoned]

SplitInfinity commented:
I'll stack them on top. I'm currently fighting some test failures due to small differences in `typing` module internals between `python3.8` and `python3.6` that impact our implementation of `is_dict`. Once I figure that out, I think those changes should transfer over to `is_list`, `is_tuple`, etc.

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

Differential Revision: [D23610766](https://our.internmc.facebook.com/intern/diff/D23610766)

[ghstack-poisoned]

SplitInfinity commented:
Re-requesting review because I had to make non-trivial changes to get `python3.6` tests to pass. Let me know what you think, and once these changes look good, I'll stack similar improvements to `try_ann_to_type` for `List`, `Tuple`, `Future`, `RRef` on top.

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -288,10 +288,23 @@ def try_ann_to_type(ann, loc):
+        else:
+            # If we don't return early here, the
+            # inspect.isclass(ann) branch below will execute
+            # in python3.6 and below, which will give the caller
```
Nice find.

OOC, what changed at 3.6 to make this not happen any more?

SplitInfinity replied:
A `Dict` annotation in python3.6 without any contained types has an `__args__` attr but it is `None`. In 3.8, that attribute is a tuple of `TypeVars`.

SplitInfinity replied:
Nvm, found a simpler solution.

Commit:
Update on "[JIT] Disallow plain Dict type annotation without arg"


**Summary**
This commit detects and prohibits the case in which `typing.Dict` is
used as an annotation without type arguments (i.e. `typing.Dict[K, V]`).
At present, `typing.Dict` is always assumed to have two arguments, and
when it is used without them, `typing.Dict.__args__` is nonempty and
contains some `typing.TypeVar` instances, which have no JIT type equivalent.
Consequently, trying to convert `typing.Dict` to a JIT type results in
a `c10::DictType` with `nullptr` for its key and value types, which can cause
a segmentation fault.

This is fixed by returning a `DictType` from
`jit.annotations.try_ann_to_type` only if the key and value types are converted
successfully to a JIT type and returning `None` otherwise.

**Test Plan**
This commit adds a unit test to `TestDict` that tests the plain `Dict`
annotations throw an error.

**Fixes**
This commit closes #43530.

Differential Revision: [D23610766](https://our.internmc.facebook.com/intern/diff/D23610766)

[ghstack-poisoned]

SplitInfinity commented:
Okay, I stacked similar PRs for `List`, `Tuple`, and `Optional` on top. `RRef` and `Future` are already handled.
